overlord/devicestate/devicestate_test.go

```diff
@@ -2817,3 +2838,446 @@ func (s *deviceMgrSuite) TestDeviceCtxProvided(c *C) {
+		Current:  siCurrent.Revision,
+		Active:   true,
+	})
+
+	tsk = st.NewTask("update-gadget", "update gadget")
+	tsk.Set("snap-setup", &snapstate.SnapSetup{
+		SideInfo: si,
+		Type:     snap.TypeGadget,
+	})
+	chg = st.NewChange("dummy", "...")
+	chg.AddTask(tsk)
+
+	st.Unlock()
+
+	return chg, tsk
+}
+
+func (s *deviceMgrSuite) TestUpdateGadgetOnCoreSimple(c *C) {
+	var updateCalled bool
+	var passedRollbackDir string
+	restore := devicestate.MockGadgetUpdate(func(current, update *gadget.Info, path string) error {
+		updateCalled = true
+		passedRollbackDir = path
+		c.Assert(osutil.IsDirectory(path), Equals, true)
+		return nil
+	})
+	defer restore()
+
+	chg, t := setupGadgetUpdate(c, s.state)
+
+	for i := 0; i < 6; i++ {
+		s.se.Ensure()
+		s.se.Wait()
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+	c.Assert(chg.IsReady(), Equals, true)
+	c.Check(chg.Err(), IsNil)
+	c.Check(t.Status(), Equals, state.DoneStatus)
+	c.Check(updateCalled, Equals, true)
+	rollbackDir := filepath.Join(dirs.SnapRollbackDir, "foo-gadget_34")
+	c.Check(rollbackDir, Equals, passedRollbackDir)
+	// should have been removed right after update
+	c.Check(osutil.IsDirectory(rollbackDir), Equals, false)
+	c.Check(s.restartRequests, DeepEquals, []state.RestartType{state.RestartSystem})
+}
+
+func (s *deviceMgrSuite) TestUpdateGadgetOnCoreNoUpdateNeeded(c *C) {
+	var called bool
+	restore := devicestate.MockGadgetUpdate(func(current, update *gadget.Info, path string) error {
+		called = true
+		return gadget.ErrNoUpdate
+	})
+	defer restore()
+
+	chg, t := setupGadgetUpdate(c, s.state)
+
+	s.se.Ensure()
+	s.se.Wait()
+
+	s.state.Lock()
+	defer s.state.Unlock()
+	c.Assert(chg.IsReady(), Equals, true)
+	c.Check(chg.Err(), IsNil)
+	c.Check(t.Status(), Equals, state.DoneStatus)
+	c.Check(called, Equals, true)
+	c.Check(s.restartRequests, HasLen, 0)
+}
+
+func (s *deviceMgrSuite) TestUpdateGadgetOnCoreDoUndo(c *C) {
+	var updateCalled bool
+	var updateRollbackDir string
+	restore := devicestate.MockGadgetUpdate(func(current, update *gadget.Info, path string) error {
+		updateRollbackDir = path
+		updateCalled = true
+		return nil
+	})
+	defer restore()
+
+	chg, t := setupGadgetUpdate(c, s.state)
+
+	s.state.Lock()
+
+	terr := s.state.NewTask("error-trigger", "provoking total undo")
+	terr.WaitFor(t)
+	chg.AddTask(terr)
+
+	s.state.Unlock()
+
+	for i := 0; i < 6; i++ {
+		s.se.Ensure()
+		s.se.Wait()
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+	c.Assert(chg.IsReady(), Equals, true)
+	c.Assert(chg.Err(), NotNil)
+	c.Check(t.Status(), Equals, state.DoneStatus)
+	c.Check(terr.Status(), Equals, state.ErrorStatus)
+	c.Check(updateCalled, Equals, true)
+	c.Check(updateRollbackDir, Equals, filepath.Join(dirs.SnapRollbackDir, "foo-gadget_34"))
+	// only one request for task do
+	c.Check(s.restartRequests, DeepEquals, []state.RestartType{state.RestartSystem})
+}
+
+func (s *deviceMgrSuite) TestUpdateGadgetOnCoreRollbackDirExistsRemovedByUpdate(c *C) {
```

Comment:
I'm not super sure what this test is testing, looking at the DeviceManager.doUpdateGadget() it seems like its checking that os.RemoveAll(snapRollbackDir) is not erroring if the path does not exists. If so, its fine to test this OTOH its a lot of test for this. Or am I misreading this?

Reply:
The idea was to make sure that we don't bail out should the update code remove rollback directory on other errors to clean up the directory.